### ISSUE TITLE
Simplify the initialisation and configuration

### ISSUE THREAD

**Issue by elitalon** [matomo-org/matomo-sdk-ios#199]
I was wondering if there's any design goal behind the need of the `configureSharedInstance` methods.

As far as I can see, the current initialisation is not expensive. And as long as `PiwikTracker` remains a singleton (related discussion on #118), we can initialise a tracker directly on the first use of `PiwikTracker.shared`.

This would have the added benefit of `PiwikTracker.shared` not being optional, which is a more common use for a singleton.

What do you think?

**Comment by brototyp** [comment]
Hey @elitalon, I am not totally sure about what you are proposing? Sure we could create the `PiwikTracker.shared` on first usage, but where would the configuration come from?

**Comment by elitalon** [comment]
I think it's confusing to have the current hybrid approach, i.e. a singleton instance that is optional and a kind-of initialiser like `configureSharedInstance`.

When you look at patterns from other libraries there are mainly two approaches:

1. Having a *non-optional* singleton instance, configured with sensible defaults and public setters for further configuration:
```swift
public final class PiwikTracker {

  public static let shared = PiwikTracker()

  internal init(…) {
  }

  public var siteUrl: URL?
}
```

2. Avoiding a singleton altogether and using a normal initializer:
```swift
public final class PiwikTracker {

  public let siteUrl: URL

  public init(siteUrl: URL) {
    self.siteUrl = siteUrl
  }

}
```

So I guess what I'm proposing is to stick to one approach instead of exposing both 😅

As a user of PiwikTracker myself, I find it annoying that `shared` is optional. However, as with any other use of the singleton pattern, keeping global state might result in unintended side effects if a client changes a property afterwards. With a normal initializer that can be avoided.

On the other hand, having a singleton object for an analytics tracker is extremely useful *in clients*. First, I don't think an app should be concerned about testing PiwikTracker. And second, passing a tracker created in the application delegate from view controller to view controller or creating a new one every time would be tedious.

**Comment by brototyp** [comment]
Thanks for your input. I think both approaches have issues in these cases.

1. I don't like the siteURL (and others) to be a variable and possibly change through the lifetime of the object. This could lead to some nasty side effects.

2. The downside of this approach is loosing the static shared variable. 

Losing the singleton isn't a big issue I think. Users of the SDK could simply extend the PiwikTracker and define their own shared instance.

```swift
extension PiwikTracker {
    public static let shared = PiwikTracker(siteId, url, ...)
}
```

This would require a little change for every user of the SDK but it isn't to bad, right?

**Comment by elitalon** [comment]
Well, stored properties are not allowed in extensions. The alternative would be to wrap PiwikTracker in another class:

```swift
final class AnalyticsTracker {
  
  static let shared  = AnalyticsTracker()

  private lazy var _piwikTracker = PiwikTracker(siteId, url,…)

  // MARK: - Initialisation

  // marked as private to go always through the singleton instance
  private init() {
  }

}
```

That's what I usually do anyway, so if the decision is to get rid of the singleton instance altogether, that's fine for me.

**Comment by brototyp** [comment]
You can't store instance properties on extension. Class or static properties a possible though. 
Thanks for your input on this. I will try that out an see if it is better this way. I think I got the general idea and I like it.

**Comment by minitrue22** [comment]
I belive singletons should be avoided as much as possible. Unfortenatly there is a tendency of overusing singletons in many iOS apps I have seen. Probably steam from the lack of a good dependency injection framework/culture.

I suggest leave the decision to the user of the framework. If the want a singleton let them create a global variable to hold an instance of an tracker. If they are more into passing around dependencies, let them do that.

**Comment by brototyp** [comment]
@elitalon Please don't close this issue. I like the discussion we are having here and I think we found a great path forward. I will reopen the issue for now.

**Comment by elitalon** [comment]
@brototyp Oops, sorry! I thought it was already clear that getting rid of `sharedInstance` was the direction to take 😅

**Comment by brototyp** [comment]
@elitalon Please have a look at the PR in #164 and let me know what you think of it.

**Comment by elitalon** [comment]
@brototyp It look goods to me overall. It's nice that the instantiation now sticks to one style.

The comment in the README regarding the possibility for clients to declare a singleton if needed is also a good addition, especially for people not aware of it.

**Comment by brototyp** [comment]
I must merged #164 which removes the shared instance and enables everyone to use more than one backend to track. It will be released in version 5.0
